bindings/go/scip/symbol_parser.go

```diff
@@ -0,0 +1,563 @@
+// prepareName, getDisambiguatorForLast} must be finished before calling
+// any of these methods again.
+func (dw *descriptorsWriter) prepareName(suffix Descriptor_Suffix) *stringWriter {
+	if dw.shouldWrite {
+		dw.tmpStringWriter.builder.Reset()
+		dw.tmpStringWriter.finalCopySlot, _ = dw.internalPrepareSlot(suffix)
+	}
+	return &dw.tmpStringWriter
+}
+
+// prepare is similar to prepareName, but allows delaying writing the suffix.
+//
+// Pre-condition: Writes to a previously written writer from {prepare,
+// prepareName, getDisambiguatorForLast} must be finished before calling
+// any of these methods again.
+func (dw *descriptorsWriter) prepare() (*stringWriter, *Descriptor_Suffix) {
+	suffixPtr := &dw.throwawaySuffix
+	if dw.shouldWrite {
+		dw.tmpStringWriter.builder.Reset()
+		dw.tmpStringWriter.finalCopySlot, suffixPtr = dw.internalPrepareSlot(Descriptor_UnspecifiedSuffix)
+	}
+	return &dw.tmpStringWriter, suffixPtr
+}
+
+// getDisambiguatorForLast returns a writer for the disambiguator for the last
+// descriptor that was inserted.
+//
+// Pre-condition: prepare or prepareName must have been called before this.
+//
+// Pre-condition: Writes to a previously written writer from {prepare,
+// prepareName, getDisambiguatorForLast} must be finished before calling
+// any of these methods again.
+func (dw *descriptorsWriter) getDisambiguatorForLast() *stringWriter {
+	if dw.shouldWrite {
+		dw.tmpStringWriter.builder.Reset()
+		dw.tmpStringWriter.finalCopySlot = &dw.descriptors[dw.count-1].Disambiguator
+		return &dw.tmpStringWriter
+	}
+	return &dw.tmpStringWriter
+}
+
+// internalPrepareSlot optionally grows the slice of descriptors, assigns the provided
+// suffix to the last descriptor, and returns pointers to the 'Name' and
+// 'Suffix' fields for overwriting.
+//
+// Generally, higher-level code should only call prepare or prepareName.
+//
+// Pre-condition: This should only be called in the shouldWrite case.
+func (dw *descriptorsWriter) internalPrepareSlot(suffix Descriptor_Suffix) (*string, *Descriptor_Suffix) {
+	assert(dw.shouldWrite, "should call internalPrepareSlot in the shouldWrite case")
+	zeroLastSlot := func() {
+		lastPtr := &dw.descriptors[dw.count]
+		if last := *lastPtr; last == nil {
+			*lastPtr = &Descriptor{Suffix: suffix}
+		} else {
+			last.Name = ""
+			last.Suffix = suffix
+			last.Disambiguator = ""
+		}
+	}
+	if len(dw.descriptors) == int(dw.count) {
+		dw.descriptors = append(dw.descriptors, &Descriptor{Suffix: suffix})
+	} else {
+		zeroLastSlot()
+	}
+	elt := dw.descriptors[dw.count]
+	dw.count++
+	nameSlot := &elt.Name
+	suffixSlot := &elt.Suffix
+	return nameSlot, suffixSlot
+}
+
+func (dw *descriptorsWriter) validDescriptors() []*Descriptor {
+	return dw.descriptors[:dw.count]
+}
+
+// symbolParserV2 parses symbols while minimizing heap allocations.
+type symbolParserV2 struct {
+	SymbolString    string
+	byteIndex       int
+	currentRune     rune
+	bytesToNextRune int32
```

Comment:
We don't expect to be constructing lots of parser objects, so I'm not concerned about potential memory usage. However, I thought it didn't make sense to have an extra branch (even if it's well predicted) in the common case to identify the length from the rune, since we already have the value computed anyways.